Add ExtUI methods to babystep probe/nozzle offsets

marcio-ao opened the pull request:
Since we are developing a dual-extruder printer, we wanted to make it as easy as possible to fine tune the Z-offsets and nozzle offsets during a print. Although UltraLCD implements babystepping for adjusting the Z-offset, the current implementation is confusing for dual-nozzles, because if the nozzle offsets are incorrect and the the user tries to adjust the Z-offset for one color to get a perfect first layer, it will be wrong for the other color. Furthermore, UltraLCD provides no ability to adjust the nozzle offsets during a print.

This PR adds a two methods to ExtUI to address the problem:

```
    void babystepAxis_mm(const float, const axis_t, bool linked_nozzles = true);
    void babystepAxis_steps(const int16_t, const axis_t, bool linked_nozzles = true);
```

These methods allow an axis to be babystepped during an active print. For a single nozzle printer, babystepping X and Y simply shifts the current print, while babystepping Z also adjusts `zprobe_zoffset`.

For the case in which the printer has multiple nozzles, the `linked_nozzles` flag indicates whether the user intends both nozzles to be adjusted together as one or whether the user intends to adjust the currently printing nozzle while keeping the other nozzles in place relative to the bed.

This combines the "Z-Offset" and "Nozzle Offsets" adjustment into a single "Adjust Offsets" screen that is active when the printer is printing:

![img_4703s](https://user-images.githubusercontent.com/29129419/49529102-96841980-f872-11e8-8c3f-be946acafb9a.JPG)

With this, a user can start a dual print, step Z up and down in order to choose the ideal "squish" for the first nozzle. Once the second color to begins printing, they can toggle the "Linked Nozzles" button to "no" and move the second nozzle relative to the material already on the bed -- i.e. step Z up and down to adjust the "squish" for that color and step X and Y to get the two colors to align perfectly on the bed.

The underlying code takes into account which nozzle is printing and adjusts either the "zprobe_offset" or the "nozzle_offset" as appropriate. Using this, the user can fully calibrate their Z-offsets and nozzle spacing from one place by making fine X, Y and Z adjustments during a dual-print.

**Caveats:**
* Since BABYSTEP_XY is not defined for delta printers, the `babystepAxis_mm` will only work for Z on deltas.
* This new functionality is only exposed via the new ExtUI. The UltraLCD implementation has not been changed.

marcio-ao committed:
Add ExtUI methods for axis babystepping

thinkyhead committed:
Update ui_api.cpp

thinkyhead commented on Marlin/src/lcd/extensible_ui/ui_api.cpp:
```diff
@@ -439,49 +439,89 @@ namespace ExtUI {
+              zprobe_zoffset += hotend_offset[Z_AXIS][0];
+              normalizeNozzleOffset(Z);
+            }
+          #endif
```
Should this just call `normalizeNozzleOffset(axis)` for all cases, since `[x][0]` or `[y][0]` could (theoretically) be changed by the loop above?

marcio-ao replied:
The reason I call `normalizeNozzleOffset(Z)` specifically in this line is because I copied `hotend_offset[Z_AXIS][0]` into `zprobe_zoffset`, so I wanted to zero out `hotend_offset[Z_AXIS][0]`. `normalizeNozzleOffset(Z);` does that for me. But as you say, I could normalize all axis if I wanted to, it's not necessary, however.

Truth be told, it's unclear whether even what I am doing here is totally correct. Come to think of it, zprobe_zoffset probably ought to be set the hotend_offset of the *lowest* nozzle.

thinkyhead replied:
Well, given that `hotend_offset[Z][0]` must definitely be 0, I think what you're doing here makes sense, and the same should also apply for `hotend_offset[X][0]` and `hotend_offset[Y][0]`.

thinkyhead replied:
> `zprobe_zoffset` probably ought to be set the `hotend_offset` of the lowest nozzle

Why would the `hotend_offset[Z]` have anything to do with the `zprobe_zoffset`? The hotend offset refers to the difference between nozzles where one might be higher or lower than the other. (The value of `hotend_offset[Z]` should always be 0 for setups where the nozzles are statically fixed to the same carriage.)

The `zprobe_zoffset` is specifically the distance between Z at the probe's trigger point and the Z at which the nozzle touches the bed. It should not be affected by operations that apply to `hotend_offset[Z]` unless the first nozzle is selected, and in that case `hotend_offset[Z]` must be zero.

marcio-ao replied:
@thinkyhead: I believe this code has been revised since, so this may not be relevent now, but originally as the user nudged the nozzles around, it could cause the hotend_offset[Z] for the first nozzle to go negative or positive. This effectively means that the user wanted the nozzle closer or further from the bed. So, in order to maintain the condition that the hotend_offset[Z] for the first nozzle was zero, I would transfer that offset to the zprobe_zoffset.

thinkyhead commented:
I like the details considered in the babystepping implementation. The joining of hotend offsets as part of babystepping makes sense. I wonder, since some babystepping behavior exists now for…
- regular LCD menu babystepping,
- the `M290` command, and
- now for ExtensibleUI

…some babystepping details are probably worth unifying under a common method, perhaps in `motion.h` / `motion.cpp`, or maybe even in a modest `Babystep` class.

marcio-ao commented:
> I like the details considered in the babystepping implementation. 

I think we may also need some clarity on what babystepping means. The current sense means introducing extra steps into an axis. This common behavior is already defined by `thermalManager.babystep_axis`, which is used by all those things you mention above. This is a useful enough thing that I think it should stand alone.

What I added was something different, it is an offset adjustment routine that makes certain assumptions about what a user may want to do when they are tuning their machines. It involves adding steps, but it does something more. It assumes that they don't necessarily just want to shift the entire print around on the bed, but that they are really trying to accomplish is to correct first layer "squish" and align nozzles on a dual color print *and* they want to save those adjustments for future prints. This entails adjusting the "zprobe_zoffset" and/or "hotend_offsets", but the idea is to hide that complexity from the user and to present them with a unified interface.

It might be clearer to continue to refer to adding steps as "babystepping", but maybe invent a new name for this real-time offset adjustment procedure. Maybe "smart offset adjustment" or "live offsets". Perhaps there is a use case for each choice:

- Babystepping - Add or remove steps (useful only to correct for shifted layers during a print).
- Smart/Live Offsets Adjustment - Add or remove steps while recording adjustments in `zoffset_zprobe` and `nozzle_offsets` to calibrate the printer for future prints.

marcio-ao commented:
> I like the details considered in the babystepping implementation. 

I think we may also need some clarity on what babystepping means. The current sense means introducing extra steps into an axis. This common behavior is already defined by `thermalManager.babystep_axis`, which is used by all those things you mention above. This is a useful enough thing that I think it should stand alone.

What I added was something different, it is an offset adjustment routine that makes certain assumptions about what a user may want to do when they are tuning their machines. It involves adding steps, but it does something more. It assumes that they don't necessarily just want to shift the entire print around on the bed, but that they are really trying to accomplish is to correct first layer "squish" and align nozzles on a dual color print *and* they want to save those adjustments for future prints. This entails adjusting the "zprobe_zoffset" and/or "hotend_offsets", but the idea is to hide that complexity from the user and to present them with a unified interface.

It might be clearer to continue to refer to adding steps as "babystepping", but maybe invent a new name for this real-time offset adjustment procedure. Maybe "smart offset adjustment" or "live offsets". Perhaps there is a use case for each choice:

- Babystepping - Add or remove steps (useful only to correct for shifted layers during a print).
- Smart/Live Offsets Adjustment - Add or remove steps while recording adjustments in `zoffset_zprobe` and `nozzle_offsets` to calibrate the printer for future prints.

thinkyhead commented:
> it is an offset adjustment routine that makes certain assumptions about what a user may want to do when they are tuning their machines.

The caveat I have with this is that it inserts expert knowledge about probe and hotend offsets into a class which shouldn't need to embody these details. It makes the most sense to keep the actual babystepping logic in `motion.*` and add behavioral options there, since babystepping is gaining common extensions —such as adjusting the `zprobe_zoffset`— as linked behavior. It would be better if it were all in one place, instead of scattered in 3 or more places.

The linkage of babystepping to other settings should be done with care. Babystepping moves an axis without changing the current position, so you wouldn't want to link babystepping to something that modified the leveling mesh or anything else which is going to get added to the logical position on its way to the planner. In the case of modifying the mesh in response to babystepping, you get a double offset — once for the babystepping, and again when the mesh is applied. Since `hotend_offset` is applied continually (on every `G1`-`G5` move), it may run into the same issue.

marcio-ao commented:
> The caveat I have with this is that it inserts expert knowledge about probe and hotend offsets into a class which shouldn't need to embody these details.

This is why I made the changes in ExtUI. It seems to me like Marlin should have some standard core functionality, but then the ExtUI API uses that core functionality to present a more user-friendly interface to the user, which in this case involves a more intuitive way to adjust offsets while printing.

It's a bit unfortunate that the UltraLCD stuff is separate as of now and it can't take advantage of the stuff in ExtUI, however.

marcio-ao committed:
Clarify method names

- "babystepping" : addition of steps to an axis to correct a layer shift.
- "smartAdjust" : tweaking an axis during a print to adjust `nozzle_offset` and `zprobe_zoffset`

marcio-ao commented:
@thinkyhead: With this latest commit, I split up the method to keep their meanings clear. "babystepAxis_steps" adds steps to an axis and is simply an ExtUI method for accessing the babystepping method in thermalManager.

The new method "smartAdjustAxis_steps" adjusts both the `zprobe_zoffset` and `nozzle_offsets` using the method described previously. I think the "smart" label makes it clear it is uses expert knowledge to change multiple values at once. The UI designer can make a choice whether to present the smart adjustment to the user, or use regular babystepping.

marcio-ao commented:
@thinkyhead: So I am really struggling to get this to work. As far as I can tell, when I call `thermalManager.babystep_axis(Z_AXIS, steps);`, the axis moves a lot less than what was requested. When I actually expect the head to move about 1mm, I need to babystep about 3mm.

I wonder if perhaps on the Archim board, the babystepping isn't working correctly. As far as I know, baby stepping happens in the temperature ISR, which is distinct from the stepper ISR. Could it be that these interrupts are interrupting each other, such that babystep pulses are getting lost?

marcio-ao commented:
I opened a ticket regarding my suspicions, issue #12635. I believe this PR is correct, but it is useless if the babystepping itself is broken and giving unreliable results.

marcio-ao commented:
@thinkyhead: See my comment on issue #12635. I have found that indeed there is a problem with the babystepping on the ISR.

marcio-ao commented:
@thinkyhead: It would be helpful if this were merged. The changes to the API work well provided that baby-stepping routines move the correct amounts (which is still an outstanding issue on some boards, but unrelated to this PR).

marcio-ao commented:
Latest push hopefully gets everything back in sync again.

thinkyhead committed:
Coding standards: correct 'if' spacing

thinkyhead commented:
Squash, rebase, tweak…
```
git fetch origin
git checkout pr-extui-adjust-axis-babysteps
git reset --hard origin/pr-extui-adjust-axis-babysteps
```
